`clippy::option_if_let_else` suggests invalid code fix

### Summary

The code produced by clippy is in theory correct in my eyes, but in reality the type checker doesn't understand it.

### Reproducer

I tried this code:

```rust
if let Some(id) = &v.as_class().unwrap().id {
    Box::new(std::iter::once(Vertex::Span(id.span)))
} else {
    Box::new(std::iter::empty())
}
```

I expected to see this happen: to get working code after a quick fix.
Instead, this happened: The quick fix produces the following non-working code:
```rust
v.as_class().unwrap().id.as_ref().map_or_else(
    || Box::new(std::iter::empty()),
    |id| Box::new(std::iter::once(Vertex::Span(id.span))),
)
```
Which is invalid because:
```
error[E0308]: mismatched types
   --> crates\oxc_linter\src\lint_adapter2.rs:217:35
    |
217 |                     |id| Box::new(std::iter::once(Vertex::Span(id.span))),
    |                          -------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Empty<_>`, found `Once<Vertex<'_>>`
    |                          |
    |                          arguments to this function are incorrect
    |
    = note: expected struct `std::iter::Empty<_>`
               found struct `std::iter::Once<lint_adapter2::Vertex<'_>>`
note: associated function defined here
   --> C:\Users\jason\.rustup\toolchains\nightly-2023-06-02-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\alloc\src\boxed.rs:215:12
    |
215 |     pub fn new(x: T) -> Self {
    |            ^^^

```

### Version

```text
rustc 1.72.0-nightly (d59363ad0 2023-06-01)
binary: rustc
commit-hash: d59363ad0b6391b7fc5bbb02c9ccf9300eef3753
commit-date: 2023-06-01
host: x86_64-pc-windows-msvc
release: 1.72.0-nightly
LLVM version: 16.0.4
```


### Additional Labels

@rustbot label +I-suggestion-causes-error

--- a/clippy_lints/src/option_if_let_else.rs
+++ b/clippy_lints/src/option_if_let_else.rs
@@ -4,8 +4,8 @@ use clippy_utils::diagnostics::span_lint_and_sugg;
 use clippy_utils::sugg::Sugg;
 use clippy_utils::ty::is_copy;
 use clippy_utils::{
-    CaptureKind, can_move_expr_to_closure, eager_or_lazy, higher, is_else_clause, is_in_const_context,
-    is_res_lang_ctor, peel_blocks, peel_hir_expr_while,
+    CaptureKind, can_move_expr_to_closure, eager_or_lazy, expr_requires_coercion, higher, is_else_clause,
+    is_in_const_context, is_res_lang_ctor, peel_blocks, peel_hir_expr_while,
 };
 use rustc_data_structures::fx::FxHashSet;
 use rustc_errors::Applicability;
@@ -210,6 +210,15 @@ fn try_get_option_occurrence<'tcx>(
             if reference_visitor.visit_expr(none_body).is_break() {
                 return None;
             }
+        }
+
+        let some_body_ty = cx.typeck_results().expr_ty(some_body);
+        let none_body_ty = cx.typeck_results().expr_ty(none_body);
+        // Check if coercion is needed for the `None` arm. If so, we cannot suggest because it will
+        // introduce a type mismatch. A special case is when both arms have the same type, then
+        // coercion is fine.
+        if some_body_ty != none_body_ty && expr_requires_coercion(cx, none_body) {
+            return None;
         }
 
         let mut app = Applicability::Unspecified;

--- a/tests/ui/option_if_let_else.fixed
+++ b/tests/ui/option_if_let_else.fixed
@@ -288,3 +288,17 @@ mod issue13964 {
         };
     }
 }
+
+mod issue11059 {
+    use std::fmt::Debug;
+
+    fn box_coercion_unsize(o: Option<i32>) -> Box<dyn Debug> {
+        if let Some(o) = o { Box::new(o) } else { Box::new("foo") }
+    }
+
+    static S: String = String::new();
+
+    fn deref_with_overload(o: Option<&str>) -> &str {
+        if let Some(o) = o { o } else { &S }
+    }
+}

--- a/tests/ui/option_if_let_else.rs
+++ b/tests/ui/option_if_let_else.rs
@@ -351,3 +351,17 @@ mod issue13964 {
         };
     }
 }
+
+mod issue11059 {
+    use std::fmt::Debug;
+
+    fn box_coercion_unsize(o: Option<i32>) -> Box<dyn Debug> {
+        if let Some(o) = o { Box::new(o) } else { Box::new("foo") }
+    }
+
+    static S: String = String::new();
+
+    fn deref_with_overload(o: Option<&str>) -> &str {
+        if let Some(o) = o { o } else { &S }
+    }
+}

